File: tasks/monitoring_prometheus.py
#!/usr/bin/env python
# Copyright 2020 Encore Technologies
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import os
import sys
import requests
import socket
import datetime
import time
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'python_task_helper', 'files'))
from task_helper import TaskHelper

class PrometheusAlertSilence(TaskHelper):

    # This will be used with the silence duration to determine how many seconds to add to the UTC end time
    def get_offset(self, units):
        return {
            'minutes': 60,
            'hours': 3600,
            'days': 86400,
            'weeks': 604800
        }.get(units, 60)

    def get_end_timestamp(self, duration, units):
        offset = self.get_offset(units)
        return datetime.datetime.utcfromtimestamp(time.time() + duration*offset).isoformat()

    # Create a silence for every target that starts now and ends after the given duration
    def create_silences(self, targets, duration, units, prometheus_server):
        silence_ids = []
        for target in targets:
            payload = {
                'matchers': [
                    {'name': 'alias', 'value': target, 'isRegex': False}
                ],
                'startsAt': datetime.datetime.utcfromtimestamp(time.time()).isoformat(),
                'endsAt': self.get_end_timestamp(duration, units),
                'comment': "Silencing alerts on {} for patching".format(target),
                'createdBy': 'patching'
            }
            res = requests.post("http://{}:9093/api/v2/silences".format(prometheus_server), json=payload)
            res.raise_for_status()
            silence_ids.append(res.json()['silenceID'])

        return silence_ids

    # Remove all silences that were created by 'patching'
    def remove_silences(self, prometheus_server):
        res = requests.get("http://{}:9093/api/v2/silences".format(prometheus_server))
        res.raise_for_status()
        silences = res.json()

        for silence in silences:
            # Remove only silences that are active and were created by 'patching'
            if silence['status']['state'] == 'active' and silence['createdBy'] == 'patching':
                requests.delete("http://{}:9093/api/v2/silence/{}".format(prometheus_server, silence['id']))
                #print(silence['id'])

    # This function either creates a silence for each name in targets or deletes all silences
    # that were created by the 'patching' user
    def task(self, args):
        duration = args['silence_duration']
        units = args['silence_units']
        targets = args['targets']
        prometheus_server = args['prometheus_server']
        action = args['action']

        ends_at = self.get_end_timestamp(duration, units)

        if action == 'disable':
            self.create_silences(targets, duration, units, prometheus_server)

        elif action == 'enable':
            self.remove_silences(prometheus_server)

if __name__ == '__main__':
    PrometheusAlertSilence().run()

File: tasks/monitoring_prometheus.rb
#!/usr/bin/env ruby
require_relative '../../ruby_task_helper/files/task_helper.rb'
require_relative '../lib/puppet_x/encore/patching/http_helper.rb'
require 'time'
require 'json'

# Bolt task for enabling/disabling monitoring alerts in SolarWinds
class MonitoringPrometheusTask < TaskHelper
  def initialize
    super()
    @http_helper = PuppetX::Patching::HTTPHelper.new(ssl: false, ssl_verify: false)
  end

  def get_end_timestamp(duration, units)
    case units
    when 'minutes'
      offset = 60
    when 'hours'
      offset = 3600
    when 'days'
      offset = 86_400
    when 'weeks'
      offset = 604_800
    end

    (Time.now.utc + duration * offset).iso8601
  end

  # Create a silence for every target that starts now and ends after the given duration
  def create_silences(targets, duration, units, prometheus_server)
    silence_ids = []
    targets.each do |target|
      payload = {
        matchers: [{ name: 'alias', value: target, isRegex: false }],
        startsAt: Time.now.utc.iso8601,
        endsAt: get_end_timestamp(duration, units),
        comment: "Silencing alerts on #{target} for patching",
        createdBy: 'patching',
      }
      headers = { 'Content-Type' => 'application/json' }
      res = @http_helper.post("http://#{prometheus_server}:9093/api/v2/silences",
                              body: payload.to_json,
                              headers: headers)

      silence_ids.push((JSON.parse res.body)['silenceID'])
    end

    silence_ids
  end

  # Remove all silences for targets that were created by 'patching'
  def remove_silences(targets, prometheus_server)
    res = @http_helper.get("http://#{prometheus_server}:9093/api/v2/silences")
    silences = res.body

    (JSON.parse silences).each do |silence|
      # Verify that the current silence is for one of the given targets
      # All silences created by this task will have exactly one matcher
      next if silence['matchers'][0]['name'] != 'alias' || !targets.include?(silence['matchers'][0]['value'])
      # Remove only silences that are active and were created by 'patching'
      if silence['status']['state'] == 'active' && silence['createdBy'] == 'patching'
        @http_helper.delete("http://#{prometheus_server}:9093/api/v2/silence/#{silence['id']}")
      end
    end
  end

  # This will either enable or disable monitoring
  def task(targets: nil,
           action: nil,
           prometheus_server: nil,
           silence_duration: nil,
           silence_units: nil,
           **_kwargs)
    # targets can be either an array or a string with a single target
    # Check if a single target was given and convert it to an array if it was
    if targets.is_a? String
      targets = [targets]
    end

    if action == 'disable'
      create_silences(targets, silence_duration, silence_units, prometheus_server)
    elsif action == 'enable'
      remove_silences(targets, prometheus_server)
    end
  end
end

MonitoringPrometheusTask.run if $PROGRAM_NAME == __FILE__
